torch/csrc/jit/pybind_utils.h

```diff
@@ -76,6 +77,69 @@ inline TypedIValue toDictKeyIValue(py::handle key) {
   }
 }
 
+inline bool isNamedTupleClass(py::object obj) {
+  auto tuple_type = reinterpret_cast<PyObject*>(&PyTuple_Type);
+  return PyObject_IsSubclass(obj.ptr(), tuple_type) &&
+      py::hasattr(obj, "_fields");
+}
+
+inline TypePtr getOrCreateNamedTupleType(
+  py::object obj,
+  const c10::QualifiedName& qualName,
+  c10::optional<std::vector<TypePtr>> element_types,
+  const SourceRange& loc) {
+  // Currently don't support default values
+  if (py::hasattr(obj, "_field_defaults")) {
+    auto default_dict = py::cast<std::map<std::string, py::object>>(
+        py::getattr(obj, "_field_defaults"));
+    if (default_dict.size()) {
+      std::string error_msg =
+          "Default values are currently not supported"
+          " on NamedTuple fields in TorchScript. Fields "
+          "with default values: [";
+      bool first = true;
+      for (const auto& kv : default_dict) {
+        if (!first) {
+          error_msg += ", ";
+        }
+        error_msg += kv.first;
+      }
+      error_msg += "]";
+      throw script::ErrorReport(loc) << error_msg;
+    }
+  }
+
+  std::string unqualName;
+  std::vector<std::string> fields;
+  c10::NamedTypePtr named_type;
+  py::object props;
+  if (element_types.has_value()) {
+    props = py::module::import("torch.jit")
+                .attr("_get_named_tuple_properties")(obj, /*need_types=*/false);
+    std::tie(unqualName, fields) = py::cast<
+        std::tuple<std::string, decltype(fields)>>(props);
+    named_type = TupleType::createNamed(qualName, fields, element_types.value());
+  } else {
+    std::vector<TypePtr> annotations;
+    props = py::module::import("torch.jit")
+                .attr("_get_named_tuple_properties")(obj, /*need_types=*/true);
+    std::tie(unqualName, fields, annotations) = py::cast<
+        std::tuple<std::string, decltype(fields), decltype(annotations)>>(props);
+    named_type = TupleType::createNamed(qualName, fields, annotations);
+  }
+
+  if (auto type = get_python_cu()->get_type(qualName)) {
+    TORCH_CHECK(
+        type->isSubtypeOf(named_type) || named_type->isSubtypeOf(type),
```

Comment:
There's another problem when trying to share the code and the Type with scripting/tracing. During tracing we infer the TensorTypes from the input and create the TensorTypes with sizes. But Scripting we only create plain TensorTypes. For the NamedTuple case, it's a container type, and NamedTupleType created from tracing might be `NamedTuple(Tensor(sizes1), Tensor(size2))`, but NamedTupleType created from scripting is just `NamedTuple(Tensor, Tensor)`. 

There's convariance problem on the subtyping relationship between these two type. I changed the subtype check here to check both conditions to ensure there's at least one pass. But ideally we need to define the behavior of it correctly. 

One idea is that when we creating the NamedTuple type from tracing, we pass the unshaped types to the type creation, that way we should the exact types between tracing/scripting. But this might affect the profiling graph as we basically erase the input tensor shapes from the named tuple.

Reply:
this get fixed by https://github.com/pytorch/pytorch/pull/30706